File: tests/test_matcher.py
"""
Unit tests for core/matcher.py
Tests word matching algorithm and annotation creation
"""

import unittest
from core.matcher import match_words, create_annotations


class TestMatchWords(unittest.TestCase):
    """Test word matching algorithm"""

    def test_all_exact_matches(self):
        """All words match exactly"""
        gt_words = ["hello", "world"]
        ocr_words = ["hello", "world"]
        matches = match_words(gt_words, ocr_words, threshold=0)

        # All should be exact matches
        exact_matches = [m for m in matches if m[3] == 'exact']
        self.assertEqual(len(exact_matches), 2)

    def test_no_matches(self):
        """No matching words"""
        gt_words = ["hello", "world"]
        ocr_words = ["foo", "bar"]
        matches = match_words(gt_words, ocr_words, threshold=0)

        # Should have 2 gt_only and 2 ocr_only
        gt_only = [m for m in matches if m[3] == 'gt_only']
        ocr_only = [m for m in matches if m[3] == 'ocr_only']
        self.assertEqual(len(gt_only), 2)
        self.assertEqual(len(ocr_only), 2)

    def test_partial_matches(self):
        """Some words match, some don't"""
        gt_words = ["hello", "world", "test"]
        ocr_words = ["hello", "foo", "test"]
        matches = match_words(gt_words, ocr_words, threshold=0)

        exact_matches = [m for m in matches if m[3] == 'exact']
        gt_only = [m for m in matches if m[3] == 'gt_only']
        ocr_only = [m for m in matches if m[3] == 'ocr_only']

        self.assertEqual(len(exact_matches), 2)  # hello, test
        self.assertEqual(len(gt_only), 1)  # world
        self.assertEqual(len(ocr_only), 1)  # foo

    def test_duplicates(self):
        """Handle duplicate words correctly"""
        gt_words = ["hello", "hello", "world"]
        ocr_words = ["hello", "world", "world"]
        matches = match_words(gt_words, ocr_words, threshold=0)

        exact_matches = [m for m in matches if m[3] == 'exact']
        gt_only = [m for m in matches if m[3] == 'gt_only']
        ocr_only = [m for m in matches if m[3] == 'ocr_only']

        # hello: match 1 (min of 2,1)
        # world: match 1 (min of 1,2)
        self.assertEqual(len(exact_matches), 2)
        self.assertEqual(len(gt_only), 1)  # one extra "hello"
        self.assertEqual(len(ocr_only), 1)  # one extra "world"

    def test_threshold_zero_no_fuzzy(self):
        """Threshold 0 should not create fuzzy matches"""
        gt_words = ["hello", "test"]
        ocr_words = ["helo", "test"]  # "helo" has edit distance 1 from "hello"
        matches = match_words(gt_words, ocr_words, threshold=0)

        fuzzy_matches = [m for m in matches if m[3] == 'fuzzy']
        self.assertEqual(len(fuzzy_matches), 0)

        # Should have 1 exact (test), 1 gt_only (hello), 1 ocr_only (helo)
        exact = [m for m in matches if m[3] == 'exact']
        gt_only = [m for m in matches if m[3] == 'gt_only']
        ocr_only = [m for m in matches if m[3] == 'ocr_only']

        self.assertEqual(len(exact), 1)
        self.assertEqual(len(gt_only), 1)
        self.assertEqual(len(ocr_only), 1)

    def test_fuzzy_matching_threshold_1(self):
        """Fuzzy matching with threshold 1"""
        gt_words = ["hello"]
        ocr_words = ["helo"]  # Edit distance 1
        matches = match_words(gt_words, ocr_words, threshold=1)

        fuzzy_matches = [m for m in matches if m[3] == 'fuzzy']
        # With threshold=1, "hello" and "helo" should fuzzy match
        self.assertEqual(len(fuzzy_matches), 1)
        self.assertEqual(fuzzy_matches[0][0], "hello")
        self.assertEqual(fuzzy_matches[0][1], "helo")
        self.assertEqual(fuzzy_matches[0][2], 1)  # edit distance

    def test_fuzzy_matching_threshold_2(self):
        """Fuzzy matching with threshold 2"""
        gt_words = ["test"]
        ocr_words = ["tst"]  # Edit distance 1
        matches = match_words(gt_words, ocr_words, threshold=2)

        fuzzy_matches = [m for m in matches if m[3] == 'fuzzy']
        self.assertEqual(len(fuzzy_matches), 1)

    def test_exact_match_priority(self):
        """Exact matches should be prioritized over fuzzy"""
        gt_words = ["hello", "world"]
        ocr_words = ["hello", "wrld"]
        matches = match_words(gt_words, ocr_words, threshold=2)

        exact = [m for m in matches if m[3] == 'exact']
        fuzzy = [m for m in matches if m[3] == 'fuzzy']

        self.assertEqual(len(exact), 1)  # "hello"
        self.assertEqual(len(fuzzy), 1)  # "world" ~ "wrld"

    def test_empty_inputs(self):
        """Handle empty inputs"""
        # Both empty
        matches = match_words([], [], threshold=1)
        self.assertEqual(len(matches), 0)

        # GT empty
        matches = match_words([], ["hello"], threshold=1)
        ocr_only = [m for m in matches if m[3] == 'ocr_only']
        self.assertEqual(len(ocr_only), 1)

        # OCR empty
        matches = match_words(["hello"], [], threshold=1)
        gt_only = [m for m in matches if m[3] == 'gt_only']
        self.assertEqual(len(gt_only), 1)

    def test_greedy_fuzzy_matching(self):
        """Greedy matching should pick best matches"""
        gt_words = ["hello", "test"]
        ocr_words = ["helo", "tst"]  # Both have distance 1
        matches = match_words(gt_words, ocr_words, threshold=1)

        fuzzy = [m for m in matches if m[3] == 'fuzzy']
        self.assertEqual(len(fuzzy), 2)

        # Check that matches are correct (not crossed)
        match_pairs = [(m[0], m[1]) for m in fuzzy]
        self.assertIn(("hello", "helo"), match_pairs)
        self.assertIn(("test", "tst"), match_pairs)


class TestCreateAnnotations(unittest.TestCase):
    """Test annotation creation for visualization"""

    def test_basic_annotations_gt(self):
        """Create annotations for ground truth"""
        word_data = [
            {'normalized': 'hello', 'original': 'Hello', 'position': 0},
            {'normalized': 'world', 'original': 'World', 'position': 1}
        ]
        matches = [
            ('hello', 'hello', 0, 'exact'),
            ('world', None, None, 'gt_only')
        ]

        annotations = create_annotations(word_data, matches, is_ground_truth=True)

        self.assertEqual(len(annotations), 2)
        self.assertEqual(annotations[0]['word'], 'Hello')
        self.assertEqual(annotations[0]['match_type'], 'exact')
        self.assertEqual(annotations[1]['word'], 'World')
        self.assertEqual(annotations[1]['match_type'], 'gt_only')

    def test_basic_annotations_ocr(self):
        """Create annotations for OCR output"""
        word_data = [
            {'normalized': 'hello', 'original': 'hello', 'position': 0},
            {'normalized': 'foo', 'original': 'foo', 'position': 1}
        ]
        matches = [
            ('hello', 'hello', 0, 'exact'),
            (None, 'foo', None, 'ocr_only')
        ]

        annotations = create_annotations(word_data, matches, is_ground_truth=False)

        self.assertEqual(len(annotations), 2)
        self.assertEqual(annotations[0]['match_type'], 'exact')
        self.assertEqual(annotations[1]['match_type'], 'ocr_only')

    def test_fuzzy_match_annotations(self):
        """Fuzzy matches in annotations"""
        word_data = [
            {'normalized': 'hello', 'original': 'hello', 'position': 0}
        ]
        matches = [
            ('hello', 'helo', 1, 'fuzzy')
        ]

        annotations = create_annotations(word_data, matches, is_ground_truth=True)

        self.assertEqual(annotations[0]['match_type'], 'fuzzy')
        self.assertEqual(annotations[0]['matched_with'], 'helo')
        self.assertEqual(annotations[0]['edit_distance'], 1)

    def test_preserves_original_word(self):
        """Annotations should use original words, not normalized"""
        word_data = [
            {'normalized': 'hello', 'original': 'HELLO', 'position': 0}
        ]
        matches = [
            ('hello', 'hello', 0, 'exact')
        ]

        annotations = create_annotations(word_data, matches, is_ground_truth=True)

        # Should preserve original case
        self.assertEqual(annotations[0]['word'], 'HELLO')

    def test_preserves_word_order(self):
        """Annotations should preserve original word order"""
        word_data = [
            {'normalized': 'one', 'original': 'One', 'position': 0},
            {'normalized': 'two', 'original': 'Two', 'position': 1},
            {'normalized': 'three', 'original': 'Three', 'position': 2}
        ]
        matches = [
            ('one', 'one', 0, 'exact'),
            ('two', None, None, 'gt_only'),
            ('three', 'three', 0, 'exact')
        ]

        annotations = create_annotations(word_data, matches, is_ground_truth=True)

        # Order should be preserved
        self.assertEqual(annotations[0]['word'], 'One')
        self.assertEqual(annotations[1]['word'], 'Two')
        self.assertEqual(annotations[2]['word'], 'Three')

    def test_duplicate_words_in_annotations(self):
        """Handle duplicate words in annotations"""
        word_data = [
            {'normalized': 'hello', 'original': 'hello', 'position': 0},
            {'normalized': 'hello', 'original': 'hello', 'position': 1}
        ]
        matches = [
            ('hello', 'hello', 0, 'exact'),
            ('hello', None, None, 'gt_only')
        ]

        annotations = create_annotations(word_data, matches, is_ground_truth=True)

        # Should handle both instances
        self.assertEqual(len(annotations), 2)
        self.assertEqual(annotations[0]['match_type'], 'exact')
        self.assertEqual(annotations[1]['match_type'], 'gt_only')


if __name__ == '__main__':
    unittest.main()

File: tests/test-matcher.js
/**
 * Test suite for matcher.js
 * Tests word matching functionality
 */

(function() {
    const tests = [];
    let passedTests = 0;
    let failedTests = 0;

    function test(name, fn) {
        tests.push({ name, fn });
    }

    function assertEquals(actual, expected, message) {
        if (JSON.stringify(actual) !== JSON.stringify(expected)) {
            throw new Error(`${message}\nExpected: ${JSON.stringify(expected)}\nActual: ${JSON.stringify(actual)}`);
        }
    }

    function assertTrue(condition, message) {
        if (!condition) {
            throw new Error(message);
        }
    }

    function runTests() {
        console.log('Running matcher.js tests...');
        passedTests = 0;
        failedTests = 0;

        for (const { name, fn } of tests) {
            try {
                fn();
                console.log(`✓ ${name}`);
                passedTests++;
            } catch (error) {
                console.error(`✗ ${name}`);
                console.error(`  ${error.message}`);
                failedTests++;
            }
        }

        console.log(`${passedTests} passed, ${failedTests} failed`);
        return { passed: passedTests, failed: failedTests, success: failedTests === 0 };
    }

    // Helper to count match types
    function countMatchTypes(matches) {
        return {
            exact: matches.filter(m => m[3] === 'exact').length,
            gt_only: matches.filter(m => m[3] === 'gt_only').length,
            ocr_only: matches.filter(m => m[3] === 'ocr_only').length
        };
    }

// Tests for matchWords

test('Match: exact matches only', () => {
    const gt = ['hello', 'world'];
    const ocr = ['hello', 'world'];
    const matches = matchWords(gt, ocr);

    const counts = countMatchTypes(matches);
    assertEquals(counts.exact, 2, 'Should have 2 exact matches');
    assertEquals(counts.gt_only, 0, 'Should have no GT-only words');
    assertEquals(counts.ocr_only, 0, 'Should have no OCR-only words');
});

test('Match: no matches', () => {
    const gt = ['hello', 'world'];
    const ocr = ['foo', 'bar'];
    const matches = matchWords(gt, ocr);

    const counts = countMatchTypes(matches);
    assertEquals(counts.exact, 0, 'Should have no exact matches');
    assertEquals(counts.gt_only, 2, 'Should have 2 GT-only words');
    assertEquals(counts.ocr_only, 2, 'Should have 2 OCR-only words');
});

test('Match: OCR missing words', () => {
    const gt = ['hello', 'world', 'test'];
    const ocr = ['hello'];
    const matches = matchWords(gt, ocr);

    const counts = countMatchTypes(matches);
    assertEquals(counts.exact, 1, 'Should have 1 exact match');
    assertEquals(counts.gt_only, 2, 'Should have 2 missing words');
});

test('Match: OCR extra words', () => {
    const gt = ['hello'];
    const ocr = ['hello', 'world', 'test'];
    const matches = matchWords(gt, ocr);

    const counts = countMatchTypes(matches);
    assertEquals(counts.exact, 1, 'Should have 1 exact match');
    assertEquals(counts.ocr_only, 2, 'Should have 2 extra words');
});

test('Match: duplicates handled correctly', () => {
    const gt = ['the', 'the', 'cat'];
    const ocr = ['the', 'cat'];
    const matches = matchWords(gt, ocr);

    const counts = countMatchTypes(matches);
    assertEquals(counts.exact, 2, 'Should match both "the" and "cat"');
    assertEquals(counts.gt_only, 1, 'Should have one unmatched "the"');
});

test('Match: non-exact words are not matched', () => {
    const gt = ['hello'];
    const ocr = ['helo']; // 1 char different
    const matches = matchWords(gt, ocr);

    const counts = countMatchTypes(matches);
    assertEquals(counts.exact, 0, 'Should have no exact matches');
    assertEquals(counts.gt_only, 1, 'Should have 1 GT-only word');
    assertEquals(counts.ocr_only, 1, 'Should have 1 OCR-only word');
});

test('Match: invoice example', () => {
    const gt = ['invoice', 'number', 'inv', '2024', '001'];
    const ocr = ['invoice', 'number', 'inv', '2o24', '001']; // O instead of 0
    const matches = matchWords(gt, ocr);

    const counts = countMatchTypes(matches);
    assertEquals(counts.exact, 4, 'Should have 4 exact matches');
    assertEquals(counts.gt_only, 1, 'Should have 1 GT-only word');
    assertEquals(counts.ocr_only, 1, 'Should have 1 OCR-only word');
});

// Tests for createAnnotations

test('Annotations: preserves original order', () => {
    const wordData = [
        { normalized: 'hello', original: 'Hello', position: 0 },
        { normalized: 'world', original: 'World!', position: 1 }
    ];
    const matches = [
        ['hello', 'hello', 0, 'exact'],
        ['world', 'world', 0, 'exact']
    ];

    const annotations = createAnnotations(wordData, matches, true);

    assertEquals(annotations[0].word, 'Hello', 'Should preserve original text');
    assertEquals(annotations[1].word, 'World!', 'Should preserve original punctuation');
});

test('Annotations: marks unmatched words', () => {
    const wordData = [
        { normalized: 'hello', original: 'Hello', position: 0 },
        { normalized: 'world', original: 'World', position: 1 }
    ];
    const matches = [
        ['hello', 'hello', 0, 'exact'],
        ['world', null, null, 'gt_only']
    ];

    const annotations = createAnnotations(wordData, matches, true);

    assertEquals(annotations[0].match_type, 'exact', 'First word should be exact');
    assertEquals(annotations[1].match_type, 'gt_only', 'Second word should be gt_only');
});

test('Annotations: assigns match IDs', () => {
    const wordData = [
        { normalized: 'hello', original: 'Hello', position: 0 }
    ];
    const matches = [
        ['hello', 'hello', 0, 'exact']
    ];

    const annotations = createAnnotations(wordData, matches, true);

    assertTrue(annotations[0].match_id !== null, 'Should have a match ID');
    assertTrue(annotations[0].match_id.startsWith('match_'), 'Match ID should have correct format');
});

test('Annotations: first-to-first matching with duplicates', () => {
    // GT has 2 'the's, OCR has 1 'the'
    // The first 'the' in GT should match with the 'the' in OCR
    const gtWordData = [
        { normalized: 'the', original: 'The', position: 0 },
        { normalized: 'cat', original: 'cat', position: 1 },
        { normalized: 'the', original: 'the', position: 2 }
    ];
    const ocrWordData = [
        { normalized: 'the', original: 'The', position: 0 },
        { normalized: 'cat', original: 'cat', position: 1 }
    ];

    const matches = matchWords(['the', 'cat', 'the'], ['the', 'cat']);

    const gtAnnotations = createAnnotations(gtWordData, matches, true);
    const ocrAnnotations = createAnnotations(ocrWordData, matches, false);

    // First 'The' in GT should have a match_id (exact match)
    assertTrue(gtAnnotations[0].match_id !== null, 'First The in GT should be matched');
    assertEquals(gtAnnotations[0].match_type, 'exact', 'First The in GT should be exact match');

    // Second 'the' in GT should NOT have a match_id (gt_only)
    assertEquals(gtAnnotations[2].match_id, null, 'Second the in GT should be unmatched');
    assertEquals(gtAnnotations[2].match_type, 'gt_only', 'Second the in GT should be gt_only');

    // The in OCR should have the same match_id as first The in GT
    assertEquals(ocrAnnotations[0].match_id, gtAnnotations[0].match_id,
        'OCR The should match with first GT The');
});

    // Register with test registry
    if (typeof window !== 'undefined' && window.TestRegistry) {
        window.TestRegistry.register('Matcher', { run: runTests });
    }

    // Export for Node.js testing
    if (typeof module !== 'undefined' && module.exports) {
        module.exports = { test, runTests, assertEquals, assertTrue };
    }
})();
